utils/scrub.1
.\" scrub - securely erase files by filling with random data first.
.\" Copyright (c) 2010 David Sugar <dyfet@gnutelephony.org>
.\"
.\" This manual page is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc.,59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
.\"
.\" This manual page is written especially for Debian GNU/Linux.
.\"
.TH scrub "1" "January 2010" "GNU uCommon" "GNU Telephony"
.SH NAME
scrub \- securely erase files by filling with random data first.
.SH SYNOPSIS
.B scrub \fI[options]\fR \fIpaths...\fR 
.br
.SH DESCRIPTION
This command is used to securely erase files.  This is accomplished by
filling the file with random data in pre-sized chunks.  Multiple passes
of random data may also be used.  The pre-sized chunks are used to remove
information about exact original file size.  Other options include random
renaming of the original file before deletion and the use of truncation to
break down meta-data on what blocks in the file system were originally 
associated with a securely deleted file.  This is specifically intended to 
make it harder to perform forensic analysis on securely erased files.
.SH OPTIONS
.TP
\fB--blocksize\fR \fIsize\fR
Set the default block size (in 1 k increments) for scrub to use when
writing random data.  This effects both the final file length, which will
be aligned to the specified size, and the way the truncate option decomposes
files.  The default is 1k.
.TP
\fB--follow\fR
Dereference and follow symlinks, erasing the target file.
.TP
\fB--passes\fR \fIcount\fR
The number of passes used when writing random data.  The default is 1 pass.
.TP
\fB--recursive\fR
If argument is a directory, recursively scan directory and any subdirectory
contents as arguments.
.TP
\fB--rename\fR
Rename the file randomly before deletion to clear persistant inode data.
.TP
\fB--truncate\fR
Decompose the file through truncation to break down file system page maps.
.TP
\fB--verbose\fR
Display each file being processed to the console.
.TP
\fB--help\fR
Outputs help screen for the user.
.SH AUTHOR
.B scrub
was written by David Sugar <dyfet@gnutelephony.org>.
.SH "REPORTING BUGS"
Report bugs to bug-commoncpp@gnu.org.
.SH COPYRIGHT
Copyright \(co 2010 David Sugar, Tycho Softworks.
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR
PURPOSE.




NEWS
uCommon Library NEWS -- history of visible user changes. Updated 2008-04-15.

uCommon C++ 0.4 codebase completed and ready for release with the introduction
of the GNU GPL V3.  This is a complete rewrite of UCommon as a C++ library for
threading and design patterns suitable for deeply embedded targets.

uCommon C++ 0.7 codebase now has core documentation and basic library theory in
header files.  This documentation can be generated with doxygen or by using
"make doxy".

uCommon C++ 1.0 release has been simplified, fully documented, and modified in
specific ways to simplify the migration and merge with GNU Common C++ to form
GNU Common C++ 2.0.

uCommon C++ 1.5 release begins the process of migrating GNU Common C++ into the
core GNU uCommon C++ codebase. Some GNU Common C++ classes, including tcp
streaming, xml parser, and persist, have been revised and integrated for the
1.5.0 release itself.  Other GNU Common C++ classes will be integrated as
requested and needed.  The 1.5/1.x series will actually be the last UCommon
release sets, as GNU uCommon C++ 2.0 will instead be released as GNU Common C++
2 2.0.

uCommon C++ 2.0 released under the GNU Lesser General Public License, Version 3
or later.

uCommon C++ 2.1 is considered a refactored abi release offering greater
clarity, consistency of use, and additional features that were migrated from
Common C++. The 2.1 release series may continue to introduce additional
classes.

ucommon 2.2 merges other ucommon related libraries to create a single ucommon
framework for application development.



utils/scrub.cpp
// Copyright (C) 2010 David Sugar, Tycho Softworks.
//
// This file is part of GNU uCommon C++.
//
// GNU uCommon C++ is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published 
// by the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// GNU uCommon C++ is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with GNU uCommon C++.  If not, see <http://www.gnu.org/licenses/>.

#include <ucommon/secure.h>
#include <sys/stat.h>

using namespace UCOMMON_NAMESPACE;

static shell::flagopt helpflag('h',"--help",	_TEXT("display this list"));
static shell::flagopt althelp('?', NULL, NULL);
static shell::numericopt blocks('b', "--blocksize", _TEXT("size of i/o blocks in k (1-x)"), "size k", 1);
static shell::flagopt follow('f', "--follow", _TEXT("follow symlinks"));
static shell::numericopt passes('p', "--passes", _TEXT("passes with randomized data (0-x)"), "count", 1);
static shell::flagopt renamefile('r', "--rename", _TEXT("rename file randomly"));
static shell::flagopt recursive('R', "--recursive", _TEXT("recursive directory scan"));
static shell::flagopt trunc('t', "--truncate", _TEXT("decompose file by truncation"));
static shell::flagopt verbose('v', "--verbose", _TEXT("show active status"));

static int exit_code = 0;
static const char *argv0 = "scrub";

static void report(const char *path, int code)
{
	const char *err = _TEXT("i/o error");

	switch(code) {
	case EACCES:
	case EPERM:
		err = _TEXT("permission denied");
		break;
	case EROFS:
		err = _TEXT("read-only file system");
		break;
	case ENODEV:
	case ENOENT:
		err = _TEXT("no such file or directory");
		break;
	case ENOTDIR:
		err = _TEXT("not a directory");
		break;
	case ENOTEMPTY:
		err = _TEXT("directory not empty");
		break;
	case ENOSPC:
		err = _TEXT("no space left on device");
		break;
	case EBADF:
	case ENAMETOOLONG:
		err = _TEXT("bad file path");
		break;
	case EBUSY:
	case EINPROGRESS:
		err = _TEXT("file or directory busy");
		break;
	case EINTR:
		err = _TEXT("operation interupted");
		break;
#ifdef	ELOOP
	case ELOOP:
		err = _TEXT("too many sym links");
		break;
#endif
	}

	if(!code) {
		if(is(verbose))
			shell::printf("%s\n", _TEXT(" removed"));
		return;
	}

	if(is(verbose))
		shell::printf(" - %s\n", err);
	else
		shell::errexit(1, "%s: %s: %s\n", argv0, path, err);

	exit_code = 1;
}

static void scrubfile(const char *path)
{
	fsys_t fs;
	struct stat ino;
	unsigned char block[1024];
	unsigned long count;
	fsys::offset_t pos = 0l;
	unsigned dots = 0;
	unsigned pass = 0;
	
	if(is(verbose))
		shell::printf("%s", path);

	int err = fsys::stat(path, &ino);

	if((err == ENOENT || fsys::islink(&ino)) && !is(follow)) {
		report(path, fsys::remove(path));
		return;
	}

	if(err == ENOENT || !ino.st_size || fsys::issys(&ino) || fsys::isdev(&ino)) {
		report(path, fsys::remove(path));
		return;
	}

	count = (ino.st_size + 1023l) / 1024;
	count /= (fsys::offset_t)(*blocks);
	count *= (fsys::offset_t)(*blocks);

	fs.open(path, fsys::ACCESS_REWRITE);
	if(!is(fs)) {
		report(path, fs.err());
		return;
	}

	while(count--) {
		while(++dots > 16384) {
			dots = 0;
			if(is(verbose))
				shell::printf(".");
		}
		pass = 0;

repeat:
		fs.seek(pos);
		if(fs.err()) {
			report(path, fs.err());
			fs.close();
			return;
		}
		if(pass < (unsigned)(*passes)) {
			Random::fill(block, 1024);
			fs.write(block, 1024);
			if(fs.err()) {
				report(path, fs.err());
				fs.close();
				return;
			}
			if(++pass < (unsigned)(*passes))
				goto repeat;
		}

		// we followup with a zero fill always as it is friendly for many 
		// virtual machine image formats which can later re-pack unused disk 
		// space, and if no random passes are specified, we at least do this.

		if(pass) 
			fs.seek(pos);

		memset(block, 0, sizeof(block));
		fs.write(block, 1024);
		if(fs.err()) {
			report(path, fs.err());
			fs.close();
			return;
		}

		pos += 1024l;
	}

	while(is(trunc) && pos > 0l) {
		pos -= 1024l * (fsys::offset_t)*blocks;
		fs.trunc(pos);
		if(fs.err()) {
			report(path, fs.err());
			fs.close();
			return;
		}
	}

	report(path, fsys::remove(path));

	fs.close();
	
}

static void scrubdir(const char *path)
{
	if(is(verbose))
		shell::printf("%s", path);
	

	report(path, fsys::removeDir(path));
}

static void dirpath(String path, bool top = true)
{
	char filename[128];
	String filepath;
	fsys_t dir(path, fsys::ACCESS_DIRECTORY);

	while(is(dir) && fsys::read(dir, filename, sizeof(filename))) {
		if(*filename == '.' && (filename[1] == '.' || !filename[1]))
			continue;

		filepath = str(path) + str("/") + str(filename);
		if(fsys::isdir(filepath)) {
			if(is(recursive))
				dirpath(filepath, false);
			else
				scrubdir(filepath);
		}
		else
			scrubfile(filepath);
	}	
	scrubdir(path);
}

extern "C" int main(int argc, char **argv)
{	
	shell::bind("scrub");
	shell args(argc, argv);
	argv0 = args.argv0();
	unsigned count = 0;

	if(*blocks < 1)
		shell::errexit(2, "%s: blocksize: %ld: %s\n",
			argv0, *blocks, _TEXT("must be greater than zero"));

	if(*passes < 0)
		shell::errexit(2, "%s: passes: %ld: %s\n",
			argv0, *passes, _TEXT("negative passes invalid"));

	argv0 = args.argv0();

	if(is(helpflag) || is(althelp)) {
		printf("%s\n", _TEXT("Usage: scrub [options] path..."));
		printf("%s\n\n", _TEXT("Echo command line arguments"));
		printf("%s\n", _TEXT("Options:"));
		shell::help();
		printf("\n%s\n", _TEXT("Report bugs to dyfet@gnu.org"));
		exit(0);
	}

	if(!args())
		return 0;

	secure::init();

	while(count < args()) {
		if(fsys::isdir(args[count]))
			dirpath(str(args[count++]));
		else
			scrubfile(args[count++]);
	}

	return exit_code;
}

